toolchain/check/import_ref.cpp

```diff
@@ -194,23 +198,30 @@ class ImportRefResolver {
       }
 
       // Step 2: resolve the instruction.
-      auto initial_work = work_stack_.size();
-      auto [new_const_id, finished] =
-          TryResolveInst(work.inst_id, existing.const_id);
-      CARBON_CHECK(finished == !HasNewWork(initial_work));
-
-      CARBON_CHECK(!existing.const_id.is_valid() ||
-                   existing.const_id == new_const_id)
-          << "Constant value changed in second pass.";
-      if (!existing.const_id.is_valid()) {
-        SetResolvedConstId(work.inst_id, existing.indirect_insts, new_const_id);
+      initial_work_ = work_stack_.size();
+      auto [new_const_id, state] = TryResolveInst(
+          work.inst_id, existing.const_id, work.abstract_const_id);
+      CARBON_CHECK((state != ResolveResult::Finished) == HasNewWork());
+
+      if (state == ResolveResult::NeedToRebuildConstant) {
```

Comment:
To note, previously the idea had been that the steps were separated out in the comments above. So to understand "Step 2", I should look under "2. Resolve the instruction". With these changes, I'm not clear that's still the case: it looks like the rebuild behavior has been broken out (actually, I'm not sure exactly where to look to understand the NeedToRebuildConstant behavior, it's not mapping well for me in the comments). You might consider shifting how you're placing comments, maybe putting them closer to where they execute.

Reply:
I've tried to make the code comments match the algorithm description a bit better.